test/extensions/filters/http/stateful_session/stateful_session_integration_test.cc

```diff
@@ -219,6 +233,151 @@ TEST_F(StatefulSessionIntegrationTest, NormalStatefulSession) {
   cleanupUpstreamAndDownstream();
 }
 
+TEST_F(StatefulSessionIntegrationTest, StatefulSessionHeaderWithPackagesAndStrictMode) {
+  initializeFilterAndRoute(
+      STATEFUL_SESSION_HEADER_FILTER_WITH_PACKAGES + STATEFUL_SESSION_STRICT_MODE, "");
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  Http::TestRequestHeaderMapImpl request_headers{{":method", "GET"},
+                                                 {":path", "/test"},
+                                                 {":scheme", "http"},
+                                                 {":authority", "stateful.session.com"}};
+
+  auto response = codec_client_->makeRequestWithBody(request_headers, 0);
+
+  auto upstream_index = waitForNextUpstreamRequest({0, 1, 2, 3});
+  ASSERT(upstream_index.has_value());
+
+  envoy::config::endpoint::v3::LbEndpoint endpoint;
+  setUpstreamAddress(upstream_index.value(), endpoint);
+  const std::string address_string =
+      fmt::format("127.0.0.1:{}", endpoint.endpoint().address().socket_address().port_value());
+  const std::string encoded_address =
+      Envoy::Base64::encode(fmt::format("{};origin:{}", address_string, "abcdefg"));
+
+  default_response_headers_.addCopy(Http::LowerCaseString("session-header"), "abcdefg");
+  upstream_request_->encodeHeaders(default_response_headers_, true);
+
+  ASSERT_TRUE(response->waitForEndStream());
+
+  EXPECT_TRUE(upstream_request_->complete());
+  EXPECT_TRUE(response->complete());
+
+  // The selected upstream server address would be selected to the response headers.
+  EXPECT_EQ(
+      encoded_address,
+      response->headers().get(Http::LowerCaseString("session-header"))[0]->value().getStringView());
+
+  cleanupUpstreamAndDownstream();
+}
+
+TEST_F(StatefulSessionIntegrationTest,
+       DownstreamRequestWithStatefulSessionHeaderWithPackagesAndStrictMode) {
+  initializeFilterAndRoute(
+      STATEFUL_SESSION_HEADER_FILTER_WITH_PACKAGES + STATEFUL_SESSION_STRICT_MODE, "");
+
+  {
+
+    envoy::config::endpoint::v3::LbEndpoint endpoint;
+    setUpstreamAddress(1, endpoint);
+    const std::string address_string =
+        fmt::format("127.0.0.1:{}", endpoint.endpoint().address().socket_address().port_value());
+    const std::string encoded_address =
+        Envoy::Base64::encode(fmt::format("{};origin:{}", address_string, "abcdefg"));
+
+    codec_client_ = makeHttpConnection(lookupPort("http"));
+    Http::TestRequestHeaderMapImpl request_headers{{":method", "GET"},
+                                                   {":path", "/test"},
+                                                   {":scheme", "http"},
+                                                   {":authority", "stateful.session.com"},
+                                                   {"session-header", encoded_address}};
```

Comment:
But here the request contains session-header and there is 'origin' part in it. Seems that Envoy passes only "abcdefg" to upstream. Can you add description why `session-header` should not be present in response?

Reply:
Because the new feature is a way to ensure the envoy could works correctly between the clients and servers which has session sticky requirements/features. So, the PACKAGES mode will only packages the original session header from server and won't add additional header if the response doesn't contains session header.

Reply:
OK. I understand it now. Thanks.